pkg/controllers/workgenerator/controller.go

```diff
@@ -0,0 +1,329 @@
+	// Gather all the resource resourceSnapshots
+	resourceSnapshots, err := r.fetchAllResourceSnapshots(ctx, resourceBinding)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// list all the corresponding works
+	works, err := r.listAllWorksAssociated(ctx, resourceBinding)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// create/update the corresponding work for each snapshot
+	activeWork := make(map[string]bool, len(resourceSnapshots))
+	for _, snapshot := range resourceSnapshots {
+		workName := getWorkNameFromSnapshotName(snapshot.Name)
+		activeWork[workName] = true
+		// TODO(RZ): issue those requests in parallel to speed up the process
+		err = r.upsertWork(ctx, works[workName], snapshot, resourceBinding)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	//  delete the works that are not associated with any resource snapshot
+	for _, work := range works {
+		if activeWork[work.Name] {
+			continue
+		}
+		klog.V(2).InfoS("Delete the work that is not associated with any resource snapshot", "work", klog.KObj(work))
+		if err := r.Client.Delete(ctx, work); err != nil {
+			klog.ErrorS(err, "failed to delete the no longer needed work", "work", klog.KObj(work))
+			if !errors.IsNotFound(err) {
+				return ctrl.Result{}, controller.NewAPIServerError(false, err)
+			}
+		}
+	}
+
+	klog.V(2).InfoS("Successfully synced all the work associated with the resourceBinding", "resourceBinding", klog.KObj(resourceBinding))
+	return ctrl.Result{}, nil
+}
+
+// fetchAllResourceSnapshots gathers all the resource snapshots for the resource binding.
+func (r *Reconciler) fetchAllResourceSnapshots(ctx context.Context, resourceBinding *fleetv1beta1.ClusterResourceBinding) (map[string]*fleetv1beta1.ClusterResourceSnapshot, error) {
+	resourceSnapshots := make(map[string]*fleetv1beta1.ClusterResourceSnapshot, 1)
+	resourceSnapshot := fleetv1beta1.ClusterResourceSnapshot{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Name: resourceBinding.Spec.ResourceSnapshotName}, &resourceSnapshot); err != nil {
+		klog.ErrorS(err, "failed to get the resource snapshot from resource resourceBinding",
+			"resource resourceBinding", klog.KObj(resourceBinding), "resource snapshot name", resourceBinding.Spec.ResourceSnapshotName)
+		return nil, controller.NewAPIServerError(true, err)
+	}
+	resourceSnapshots[resourceSnapshot.Name] = &resourceSnapshot
+
+	// check if there are more snapshot in the same index group
+	countAnnotation := resourceSnapshot.Annotations[fleetv1beta1.NumberOfResourceSnapshotsAnnotation]
+	snapshotCount, err := strconv.Atoi(countAnnotation)
+	if err != nil {
+		return nil, controller.NewUnexpectedBehaviorError(err)
+	}
+	if snapshotCount > 1 {
+		index, exist := resourceSnapshot.Labels[fleetv1beta1.ResourceIndexLabel]
+		if !exist {
+			unexpectedErr := fmt.Errorf("resource snapshot %s has no index label", resourceSnapshot.Name)
```

Comment:
add log?

Reply:
The reason I didn't was that the framework will print out a log at the end of any failed reconcile so it's kinda a dup